core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -547,19 +536,30 @@ class TransactionStateManager(brokerId: Int,
             val metadata = epochAndTxnMetadata.transactionMetadata
             metadata synchronized {
               if (epochAndTxnMetadata.coordinatorEpoch == coordinatorEpoch) {
-                debug(s"TransactionalId ${metadata.transactionalId}, resetting pending state since we are returning error $responseError")
-                metadata.pendingState = None
+                if (retryOnError(responseError)) {
```

Comment:
Just did.

I found another case where the `txn state manager` append call could return an `loading in progress` error, which will fall into `other` case to throw an exception, I have added this along with the commit as well.